src/loaders/lottie/tvgLottieBuilder.cpp

```diff
@@ -1083,42 +1083,67 @@ void LottieBuilder::updateText(LottieLayer* layer, float frameNo)
                     shape->strokeFill(doc.stroke.color.rgb[0], doc.stroke.color.rgb[1], doc.stroke.color.rgb[2]);
                 }
 
-                bool needGroup = false;
+                auto needGroup = false;
+                //text range process
                 if (!text->ranges.empty()) {
                     Point scaling = {1.0f, 1.0f};
                     auto rotation = 0.0f;
                     Point translation = {0.0f, 0.0f};
+                    auto setColor = false;
 
-                    //text range process
                     for (auto s = text->ranges.begin(); s < text->ranges.end(); ++s) {
-                        float start, end;
-                        (*s)->range(frameNo, float(totalChars), start, end);
-
                         auto basedIdx = idx;
                         if ((*s)->based == LottieTextRange::Based::CharsExcludingSpaces) basedIdx = idx - space;
                         else if ((*s)->based == LottieTextRange::Based::Words) basedIdx = line + space;
                         else if ((*s)->based == LottieTextRange::Based::Lines) basedIdx = line;
 
-                        if (basedIdx < start || basedIdx >= end) continue;
+                        auto f = (*s)->factor(frameNo, float(totalChars), (float)basedIdx);
+                        if (tvg::zero(f)) continue;
                         needGroup = true;
 
-                        translation = translation + (*s)->style.position(frameNo);
-                        scaling = scaling * (*s)->style.scale(frameNo) * 0.01f;
-                        rotation += (*s)->style.rotation(frameNo);
+                        translation = translation + f * (*s)->style.position(frameNo);
+                        scaling = scaling * (f * ((*s)->style.scale(frameNo) * 0.01f - Point{1.0f,1.0f}) + Point{1.0f,1.0f});
+                        rotation += f * (*s)->style.rotation(frameNo);
 
                         shape->opacity((*s)->style.opacity(frameNo));
 
-                        auto color = (*s)->style.fillColor(frameNo);
-                        shape->fill(color.rgb[0], color.rgb[1], color.rgb[2], (*s)->style.fillOpacity(frameNo));
+                        auto rangeColor = (*s)->style.fillColor(frameNo);
+                        if (tvg::equal(f, 1.0f)) shape->fill(rangeColor.rgb[0], rangeColor.rgb[1], rangeColor.rgb[2], (*s)->style.fillOpacity(frameNo));
+                        else {
+                            auto color = doc.color;
```

Comment:
1.  What about opacity?

2. Suggestion logic:
```
auto color = doc.color;
auto strokeColor = doc.stroke.color;
for (auto s = text->ranges.begin(); s < text->ranges.end(); ++s) {
   ...
   color = interpolate (color, rangeColor, f);
   ...
}
shape->fill(color);
if (doc.stroke.render) shape->strokeFill(strokeColor);
```

Reply:
sample for opacity - works
[opacity.json](https://github.com/user-attachments/files/17633984/opacity.json)

Reply:
@mgrudzinska We cannot confirm with this sample, as it always has a factor of 1.0f, which means opacity with interpolation also works as expected.

Reply:
By default, interpolating other properties, including color, while excluding only opacity does not seem to align with lottie spec. It would be better to validate cases where issues arise when interpolating opacity.

Reply:
By default, interpolating other properties, including color, while excluding only opacity does not seem to align with lottie spec. It would be better to suggest cases where issues arise when interpolating opacity.

Reply:
@hermet please check #2949